client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -248,33 +248,27 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = {
-          val entity = req.entity match {
-            case Entity.Empty | Entity.Strict(_) =>
-              req
-            case Entity.Default(_, _) =>
-              req.pipeBodyThrough(until(disposed))
-          }
-
-          addHostHeaderIfUriIsAbsolute(entity)
+    def runDefault(req: Request[F]) =
+      Resource.suspend(
+        Ref[F].of(false).map { disposed =>
+          val req0 =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(untilDisposed(disposed)))
+          Resource
+            .eval(app(req0))
+            .onFinalize(disposed.set(true))
+            .map(_.pipeBodyThrough(untilDisposed(disposed)))
```

Comment:
@danicheg is it OK that I removed the entity check on the Response here?  I'm not sure if a Request has an Entity type of Default if the Response must also always be Default.

Reply:
This code has problems.  If the request body isn't a Stream the response could be a Stream.